.github/workflows/build.yml

```diff
@@ -129,13 +134,80 @@ jobs:
+        wget -qO- https://apt.llvm.org/llvm-snapshot.gpg.key | sudo tee /etc/apt/trusted.gpg.d/apt.llvm.org.asc
+        DISTRO_CODENAME="$(lsb_release --codename --short)"
+        sudo add-apt-repository -y "deb http://apt.llvm.org/$DISTRO_CODENAME/ llvm-toolchain-$DISTRO_CODENAME-18 main"
+        sudo apt update -y
+        sudo apt install -y clang-18
+      if: matrix.name == 'debug' && startsWith(matrix.os, 'ubuntu')
+    - name: Set global configure flags (debug asan)
+      run: |
+        # These flags get applied to both the Ruby compilation _and_ the compilation of dependencies
+        # wiht ASAN enabled
+        echo "optflags=-O3 -fno-inline -fno-omit-frame-pointer" >> $GITHUB_ENV
+        # Compile with ASAN
+        echo "debugflags=-fsanitize=address -ggdb3 -fPIC" >> $GITHUB_ENV
+        # Requires clang 18
+        echo "CC=clang-18" >> $GITHUB_ENV
+      if: matrix.name == 'debug' && startsWith(matrix.os, 'ubuntu')
+    - name: Build ASAN dependencies (debug asan)
+      run: |
+        set -ex
+        # Where we're going to store ASAN static archives (these need to be distributed, since the LDFLAGS
+        # we need to use to link Ruby against them get burned into the resulting rbconfig)
+        mkdir -p "$RUBY_PREFIX/asan"
+        mkdir asan_build
+        pushd asan_build
+
+        OPENSSL_VERSION="3.3.0"
+        LIBYAML_VERSION="0.2.5"
+
+        # Make sure we compile OpenSSL to look for certificates in the same place that the
+        # distribution provided OpenSSl would.
+        OPENSSLDIR="$(openssl version -d | ruby -rshellwords -ne 'if $_ =~ /OPENSSLDIR:(.*)$/; puts Shellwords.shellsplit($1.strip); end')"
+
+        curl -fsSLO "https://www.openssl.org/source/openssl-${OPENSSL_VERSION}.tar.gz"
+        tar xf "openssl-${OPENSSL_VERSION}.tar.gz"
+        pushd "openssl-${OPENSSL_VERSION}"
+        ./Configure --prefix="$RUBY_PREFIX/asan" --libdir=lib --openssldir="$OPENSSLDIR" no-shared no-tests no-apps CC="$CC" CFLAGS="$optflags $debugflags"
+        make -j4
+        # make install_sw will try and write config to OPENSSLDIR, which we don't want to do.
+        make install_dev
+        popd
+
+        curl -fsSLO "http://pyyaml.org/download/libyaml/yaml-${LIBYAML_VERSION}.tar.gz"
+        tar xf "yaml-${LIBYAML_VERSION}.tar.gz"
+        pushd "yaml-${LIBYAML_VERSION}"
+        ./configure --prefix="$RUBY_PREFIX/asan" --disable-shared --enable-static CC="$CC" CFLAGS="$optflags $debugflags"
+        make -j4
+        make install
+        popd
+
+        popd
+        rm -Rf asan_build
+      if: matrix.name == 'debug' && startsWith(matrix.os, 'ubuntu')
+    - name: Set Ruby configure flags (debug asan)
+      run: |
+        # These flags get applied only to the Ruby build, not to the libs.
+        # Clang > 17 does not work with M:N threading: https://bugs.ruby-lang.org/issues/20243
+        echo "cppflags=-DENABLE_PATH_CHECK=0 -DRUBY_DEBUG=1 -DVM_CHECK_MODE=1 -DUSE_MN_THREADS=0" >> $GITHUB_ENV
```

Comment:
This is some duplication + override and it's far from the previous usage, that seems problematic.
I think it could be addressed by moving the other configure command lines closer to here.